invenio_mail/tasks.py

```diff
@@ -79,4 +85,40 @@ def _send_email_with_attachments(data, attachments=[]):
             disposition = attachment.get("disposition")
         msg.attach(content_type=content_type, data=rawdata, disposition=disposition)
 
-    current_app.extensions["mail"].send(msg)
+    _send(msg, self)
+
+
+def _send(msg, task):
+    """Send emails and add exception handling."""
+    try:
+        current_app.extensions["mail"].send(msg)
+    except (
+        smtplib.SMTPRecipientsRefused,
+        smtplib.SMTPHeloError,
+        smtplib.SMTPSenderRefused,
+        smtplib.SMTPDataError,
+        smtplib.SMTPNotSupportedError,
+    ):
+        if task.request.retries <= current_app.config["MAIL_MAX_RETRIES"]:
+            raise task.retry(
+                countdown=int(random.uniform(2, 4) ** task.request.retries)
+            )
+        else:
+            current_app.logger.error("Mail could not be dispatched!")
+            if current_app.config["MAIL_LOG_FAILED_MESSAGES"]:
+                current_app.logger.info(msg.body)
+                if msg.attachments:
+                    for attachment in msg.attachments:
+                        current_app.logger.info(attachment.data)
+    except Exception:
+        if task.request.retries <= current_app.config["MAIL_MAX_RETRIES"]:
+            raise task.retry(
+                countdown=int(random.uniform(2, 4) ** task.request.retries)
+            )
+        else:
+            current_app.logger.error("Mail server does not answer to requests!")
+            if current_app.config["MAIL_LOG_FAILED_MESSAGES"]:
+                current_app.logger.info(msg.body)
+                if msg.attachments:
+                    for attachment in msg.attachments:
+                        current_app.logger.info(attachment.data)
```

Comment:
Looks like the only difference is the msg?

```python

    try:
        current_app.extensions["mail"].send(msg)
    except (...SMTP ones, Exception) as excp:
        msg = "...could not be dispatched" if instanceof(excp, SMTPException) else "... does not answer"
        ....
```
or call a common func instead, if SMTPException?

Reply:
Yes, it's much cleaer and shorter w/o duplicated code :-)